components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java

```diff
@@ -2147,29 +2236,89 @@ public void updateApplication(Application application) throws APIManagementExcep
-                getAPIManagerConfigurationService().getAPIManagerConfiguration();
+        boolean isTenantFlowStarted = false;
+        int applicationId = application.getId();
+
+        if (tenantDomain != null && !MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+            PrivilegedCarbonContext.startTenantFlow();
+            isTenantFlowStarted = true;
+            PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain, true);
+        }
 
         try {
-            //Remove from cache first since we won't be able to find active access tokens once the application is removed.
+            String workflowExtRef = null;
+            WorkflowExecutor createApplicationWFExecutor = WorkflowExecutorFactory.getInstance().
+                    getWorkflowExecutor(WorkflowConstants.WF_TYPE_AM_APPLICATION_CREATION);
+            WorkflowExecutor createSubscriptionWFExecutor = WorkflowExecutorFactory.getInstance().
+                    getWorkflowExecutor(WorkflowConstants.WF_TYPE_AM_SUBSCRIPTION_CREATION);
+            WorkflowExecutor createProductionRegistrationWFExecutor = WorkflowExecutorFactory.getInstance().
+                    getWorkflowExecutor(WorkflowConstants.WF_TYPE_AM_APPLICATION_REGISTRATION_PRODUCTION);
+            WorkflowExecutor createSandboxRegistrationWFExecutor = WorkflowExecutorFactory.getInstance().
+                    getWorkflowExecutor(WorkflowConstants.WF_TYPE_AM_APPLICATION_REGISTRATION_SANDBOX);
+            WorkflowExecutor removeApplicationWFExecutor = WorkflowExecutorFactory.getInstance().
+                    getWorkflowExecutor(WorkflowConstants.WF_TYPE_AM_APPLICATION_DELETION);
+
+            workflowExtRef = apiMgtDAO.getExternalWorkflowReferenceByApplicationID(application.getId());
+
+            ApplicationWorkflowDTO workflowDTO = new ApplicationWorkflowDTO();
+            workflowDTO.setApplication(application);
+            workflowDTO.setWorkflowType(WorkflowConstants.WF_TYPE_AM_APPLICATION_DELETION);
+            workflowDTO.setTenantDomain(tenantDomain);
+            workflowDTO.setTenantId(tenantId);
+            workflowDTO.setExternalWorkflowReference(workflowExtRef);
+
+            // Remove from cache first since we won't be able to find active access tokens
+            // once the application is removed.
             invalidateCachedKeys(application.getId());
-        }catch(APIManagementException ignore){
-            //Log and proceed since we do not want to halt the application delete due to cache invalidation failures.
-            log.warn("Failed to invalidate Gateway Cache " + ignore.getMessage());
+
+            // clean up pending subscription tasks
+            Set<Integer> pendingSubscriptions = apiMgtDAO.getPendingSubscriptionsByApplicationId(applicationId);
+            for (int subscription : pendingSubscriptions) {
+                workflowExtRef = apiMgtDAO.getExternalWorkflowReferenceForSubscription(subscription);
```

Comment:
Shouldn't we catch any exception that is been thrown when cleaning the pending tasks so that the loop will not break if there is any exception when cleaning up a pending task